minion/etcd/network.go

```
@@ -27,290 +31,354 @@ const gatewayIP = "10.0.0.1"
 func wakeChan(conn db.Conn, store Store) chan struct{} {
-	labelWatch := store.Watch(labelDir, 1*time.Second)
-	containerWatch := store.Watch(labelDir, 1*time.Second)
+	minionWatch := store.Watch(minionDir, 1*time.Second)
 	trigg := conn.TriggerTick(30, db.MinionTable, db.ContainerTable, db.LabelTable,
 		db.EtcdTable).C
 
 	c := make(chan struct{}, 1)
 	go func() {
 		for {
 			select {
-			case <-labelWatch:
-			case <-containerWatch:
+			case <-minionWatch:
 			case <-trigg:
 			}
 
 			select {
 			case c <- struct{}{}:
-			default: // There's a notification in queue no need for another.
+			default: // There's a notification in queue, no need for another.
 			}
 		}
 	}()
 
 	return c
 }
 
-func runNetworkWorker(conn db.Conn, store Store) {
-	// If the directories don't exist, create them so we may watch them.  If they
-	// exist already these will return an error that we won't log, but that's ok
-	// cause the loop will error too.
-	store.Mkdir(labelDir)
-	store.Mkdir(containerDir)
+func runNetwork(conn db.Conn, store Store) {
+	if err := store.Mkdir(minionDir); err != nil {
+		log.WithError(err).Info()
+	}
 
 	for range wakeChan(conn, store) {
-		labelDir, err := getDirectory(store, labelDir)
-		containerDir, err2 := getDirectory(store, containerDir)
-		if err2 != nil {
-			err = err2
-		}
-
-		if err != nil {
-			log.WithError(err).Warn("Failed to read from cluster store.")
-			continue
-		}
+		leader := false
+		var containers []db.Container
+		etcdData, err := readEtcd(store)
 
 		conn.Transact(func(view db.Database) error {
-			readContainerTransact(view, containerDir)
-			readLabelTransact(view, labelDir)
+			leader = view.EtcdLeader()
+			containers = view.SelectFromContainer(func(c db.Container) bool {
+				return c.DockerID != ""
+			})
+
+			// The leader updates both the store and the etcdData even
+			// if the read fails
+			if leader {
+				etcdData = updateEtcd(store, etcdData, containers)
+			}
+
+			if err != nil {
```

Yeah, as noted with the above point.  Unless theirs a really good reason, we can just return an error when the read fails, and move this error check up out of the transaction.